lib/matplotlib/cm.py

```diff
@@ -143,7 +152,17 @@ def get_cmap(name=None, lut=None):
         return name
     cbook._check_in_list(sorted(cmap_d), name=name)
     if lut is None:
+        if cmap_d[name]._global_changed:
+            cbook.warn_deprecated(
+                "3.3",
+                message="The colormap requested has had the global state "
+                        "changed without being registered. Accessing a "
+                        "colormap in this way has been deprecated. "
+                        "Please register the colormap using "
+                        "plt.register_cmap() before requesting "
+                        "the modified colormap.")
```

Comment:
I understand this, but only because I read the thread.  If I was a user who ran 
```python
plt.get_cmap('viridis').set_bad('k')
fig, ax = plt.subplots()
ax.imshow(data, cmap='viridis')
```
I'd have no idea what I was supposed to do here.  I also don't think we should be encouraging any changing of the global state.  I'd far rather tell folks that they should deal with the camp object which is far less arcane than registering the global state:

```python
cmp = plt.get_cmap('viridis').set_bad('k')
fig, ax = plt.subplots()
ax.imshow(data, cmap=cmp)
```

To that end, I'd say "Modifying a registered colormap is deprecated; make a new copy via cmp=plt.get_cmap('viridian') and modify the copy." ?

Reply:
Unfortunately, that won't even work either! (which is why I think this is a bug) You would need to cmp=copy.copy(plt.get_cmap('viridian')) currently to get the desired functionality. The real problem is that further down if you go and request 'viridis' from a plot after doing a bunch of stuff, you will still get the bad colormap.